src/CMakeLists.txt

```diff
@@ -83,121 +83,69 @@ set(COORDINATE_TYPE UniformCartesian) # TODO: Make this an option when more are
-  outputs/parthenon_hdf5.cpp
-  outputs/restart.cpp
-  outputs/vtk.cpp
-
-  parthenon/driver.hpp
-  parthenon/package.hpp
-  parthenon/parthenon.hpp
-  parthenon/prelude.hpp
-
-  pgen/default_pgen.cpp
-
-  reconstruct/characteristic.cpp
-  reconstruct/dc_inline.hpp
-  reconstruct/dc.cpp
-  reconstruct/plm_inline.hpp
-  reconstruct/plm.cpp
-  reconstruct/ppm.cpp
-  reconstruct/reconstruction.hpp
-  reconstruct/reconstruction.cpp
-
-  refinement/amr_criteria.cpp
-  refinement/amr_criteria.hpp
-  refinement/refinement.cpp
-  refinement/refinement.hpp
-
-  tasks/task_id.cpp
-  tasks/task_id.hpp
-  tasks/task_list.hpp
-  tasks/task_types.hpp
-
-  utils/buffer_utils.cpp
-  utils/buffer_utils.hpp
-  utils/change_rundir.cpp
-  utils/error_checking.hpp
-  utils/show_config.cpp
-  utils/signal_handler.cpp
-  utils/trim_string.cpp
-  utils/trim_string.hpp
-  utils/utils.hpp
-
-  argument_parser.hpp
-  basic_types.hpp
-  defs.hpp
-  globals.cpp
-  globals.hpp
-  kokkos_abstraction.hpp
-  parameter_input.cpp
-  parameter_input.hpp
-  parthenon_manager.cpp
-  parthenon_manager.hpp
-  parthenon_mpi.hpp
-)
+file( GLOB base_src *.cpp )
+file( GLOB base_hdr *.hpp )
+file( GLOB bvals_src bvals/*.cpp bvals/fc/*.cpp bvals/cc/*.cpp)
+file( GLOB bvals_hdr bvals/*.hpp bvals/fc/*.hpp bvals/cc/*.hpp)
+file( GLOB coordinates_src coordinates/*.cpp )
+file( GLOB coordinates_hdr coordinates/*.hpp )
+file( GLOB driver_src driver/*.cpp )
+file( GLOB driver_hdr driver/*.hpp )
+file( GLOB interface_src interface/*.cpp )
+file( GLOB interface_hdr interface/*.hpp )
+file( GLOB mesh_src mesh/*.cpp )
+file( GLOB mesh_hdr mesh/*.hpp )
+file( GLOB outputs_src outputs/*.cpp )
+file( GLOB outputs_hdr outputs/*.hpp )
+file( GLOB parthenon_src parthenon/*.cpp )
+file( GLOB parthenon_hdr parthenon/*.hpp )
+file( GLOB pgen_src pgen/*.cpp )
+file( GLOB pgen_hdr pgen/*.hpp )
+file( GLOB reconstruct_src reconstruct/*.cpp )
+file( GLOB reconstruct_hdr reconstruct/*.hpp )
+file( GLOB refinement_src refinement/*.cpp )
+file( GLOB refinement_hdr refinement/*.hpp )
+file( GLOB tasks_src tasks/*.cpp )
+file( GLOB tasks_hdr tasks/*.hpp )
+file( GLOB utils_src utils/*.cpp )
+file( GLOB utils_hdr utils/*.hpp )
+
+list(APPEND parthenon_subdirs
+  "bvals"
+  "coordinates"
+  "driver"
+  "interface"
+  "mesh"
+  "outputs"
+  "parthenon"
+  "pgen"
+  "reconstruct"
+  "refinement"
+  "tasks"
+  "utils")
+
+list(APPEND parthenon_sources
+     "${bvals_src}"
+     "${coordinates_src}"
+     "${driver_src}"
+     "${interface_src}"
+     "${mesh_src}"
+     "${outputs_src}"
+     "${parthenon_src}"
+     "${pgen_src}"
+     "${reconstruct_src}"
+     "${refinement_src}"
+     "${tasks_src}"
+     "${utils_src}"
+     "${base_src}"
+     )
+
+add_library(parthenon SHARED "${parthenon_sources}")
+
+# Other headers will be installed manually later to maintain directory structure
+list(APPEND public_headers "${base_hdr}")
+set_target_properties(parthenon PROPERTIES PUBLIC_HEADER "${public_headers}")
```

Comment:
This feels to me like an improvement over what we had before (and I also understand it's now necessary to gather the headers). However, I recall @AndrewGaspar expressing some discomfort with using `glob` to gather source files. I don't understand the concern fully though. It'd be interesting to hear the cmake wizards among us chime in. @JoshuaSBrown do you know?

If we are going to go a glob route, could we go further? Could we automatically generate the directory structure, rather than having to enter it by hand? If I recall, @JoshuaSBrown and @AndrewGaspar did something like this for the testing parts of the code.

Reply:
I'm sure there are more elegant ways to handle this than what I've done above. And `GLOB` may be restrictive in some cases. It *is* very convenient to use a cmake call to copy all `*.hpp` files recursively into the install directory with the same data structure though.

Reply:
It is a bad idea to use globs in this case, here is a quote from cmake:

> [Note: We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate. The CONFIGURE_DEPENDS flag may not work reliably on all generators, or if a new generator is added in the future that cannot support it, projects using it will be stuck. Even if CONFIGURE_DEPENDS works reliably, there is still a cost to perform the check on every rebuild.](https://cmake.org/cmake/help/v3.15/command/file.html?highlight=glob#file)

This has been the source of headaches/wasted time for many developers. There are some cases where globbing makes sense I don't think this is one of them. Granted it is a lot cleaner, however problems come any time there are random files lying around. Let me give you an example you are building on a git branch and you create a file but forget to add it to the branch, you then switch branches and try to build, you will probably end up with a linking error.

Some of these scenarios are easy to debug but others are not.

Reply:
I didn't realize `GLOB` was so dangerous! I'll get rid of those calls.

Reply:
I personally don't feel that GLOB is actually all the dangerous, especially when used with `CONFIGURE_DEPENDS`, but we've already committed to not using it, so please don't change it.

Reply:
I originally favored glob as a consequence of how I was manually building up a directory structure to manually handle header installs among other things, but it's no longer necessary now that I better understand `install` syntax so all these GLOBs are gone and I've reverted to nearly the original syntax for listing source/header files.